algorithms/kernel/dtrees/forest/regression/oneapi/df_regression_train_dense_default_oneapi_impl.i

```diff
@@ -0,0 +1,1245 @@
+
+    {
+        KernelArguments args(1);
+        args.set(0, treeOrder, AccessModeIds::write);
+
+        KernelRange global_range(nRows);
+
+        context.run(global_range, kernel, args, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+    }
+
+    return status;
+}
+
+template <typename algorithmFPType, decision_forest::regression::training::Method method>
+services::Status RegressionTrainBatchKernelOneAPI<algorithmFPType, method>::getOOBRows(const UniversalBuffer & rowsList, size_t nRows,
+                                                                                       size_t & nOOBRows, UniversalBuffer & oobRowsList)
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(compute.getOOBRowsList);
+
+    services::Status status;
+
+    auto & context = services::Environment::getInstance()->getDefaultExecutionContext();
+
+    const int absentMark    = -1;
+    const int localSize     = _preferableSubGroup;
+    const int nSubgroupSums = _maxLocalSums * localSize < nRows ? _maxLocalSums : (nRows / localSize + !(nRows / localSize));
+
+    auto rowsBuffer        = context.allocate(TypeIds::id<int>(), nRows, &status); // it is filled with marks Present/Absent for each rows
+    auto partialSums       = context.allocate(TypeIds::id<int>(), nSubgroupSums, &status);
+    auto partialPrefixSums = context.allocate(TypeIds::id<int>(), nSubgroupSums, &status);
+    auto totalSum          = context.allocate(TypeIds::id<int>(), 1, &status);
+    DAAL_CHECK_STATUS_VAR(status);
+
+    context.fill(rowsBuffer, absentMark, &status);
+    DAAL_CHECK_STATUS_VAR(status);
+
+#ifdef CHECKTIME
+    timer mark_st = timer::start();
+#endif
+
+    {
+        auto & kernel = kernelMarkPresentRows;
+        KernelArguments args(3);
+        args.set(0, rowsList, AccessModeIds::read);
+        args.set(1, rowsBuffer, AccessModeIds::write);
+        args.set(2, (int)nRows);
+
+        KernelRange local_range(localSize);
+        KernelRange global_range(localSize * nSubgroupSums);
+
+        KernelNDRange range(1);
+        range.local(local_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+        range.global(global_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        context.run(range, kernel, args, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+    }
+
+    {
+        auto & kernel = kernelCountAbsentRowsForBlocks;
+        KernelArguments args(3);
+        args.set(0, rowsBuffer, AccessModeIds::read);
+        args.set(1, partialSums, AccessModeIds::write);
+        args.set(2, (int)nRows);
+
+        KernelRange local_range(localSize);
+        KernelRange global_range(localSize * nSubgroupSums);
+
+        KernelNDRange range(1);
+        range.local(local_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+        range.global(global_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        context.run(range, kernel, args, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+    }
+
+    {
```

Comment:
these three functions are run one after another. Could it be carried out in a separate function?

Reply:
Yes, it is generally better to move actual kernel execution (with necessary preprocessing, like range creation, etc.) into a separate functions with meaningful names.
This will help to understand the flow of this big function.

Reply:
moved into separate methods